src/rng/src/vhu_rng.rs

```diff
@@ -0,0 +1,263 @@
+use virtio_bindings::bindings::virtio_net::*;
+use virtio_bindings::bindings::virtio_ring::{
+    VIRTIO_RING_F_EVENT_IDX, VIRTIO_RING_F_INDIRECT_DESC,
+};
+use vm_memory::{
+    Bytes, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap
+};
+use vmm_sys_util::eventfd::EventFd;
+
+const QUEUE_SIZE: usize = 1024;
+const NUM_QUEUES: usize = 1;
+
+type Result<T> = std::result::Result<T, Error>;
+type VhostUserBackendResult<T> = std::result::Result<T, std::io::Error>;
+
+#[derive(Debug)]
+/// Errors related to vhost-device-i2c daemon.
+pub enum Error {
+    /// Failed to create kill eventfd.
+    CreateKillEventFd(io::Error),
+    /// Failed to handle event other than input event.
+    HandleEventNotEpollIn,
+    /// Failed to handle unknown event.
+    HandleEventUnknownEvent,
+    /// Guest gave us a readable descriptor that protocol says to only write to.
+    UnexpectedReadDescriptor,
+    /// No memory configured.
+    NoMemoryConfigured,
+    /// Failed to read from the RNG source.
+    UnexpectedRngSourceError,
+    /// Previous Time value is later than current time.
+    UnexpectedTimerValue,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "virtiorng_error: {:?}", self)
+    }
+}
+
+impl error::Error for Error {}
+
+impl convert::From<Error> for io::Error {
+    fn from(e: Error) -> Self {
+        io::Error::new(io::ErrorKind::Other, e)
+    }
+}
+
+pub struct VuRngTimerConfig {
+    period_ms: u128,
+    period_start: Instant,
+    max_bytes: usize,
+    quota_remaining: usize,
+}
+
+impl VuRngTimerConfig {
+    pub fn new(period_ms: u128, max_bytes: usize) -> Self {
+        VuRngTimerConfig {
+                        period_ms,
+                        period_start: Instant::now(),
+                        max_bytes,
+                        quota_remaining: max_bytes,
+        }
+    }
+}
+
+pub struct VuRngBackend {
+    mem: Option<GuestMemoryAtomic<GuestMemoryMmap>>,
+    event_idx: bool,
+    timer: Mutex<VuRngTimerConfig>,
+    random_file: Mutex<File>,
+    pub exit_event: EventFd,
+}
+
+impl VuRngBackend {
+    /// Create a new virtio rng device that gets random data from /dev/urandom.
+    pub fn new(
+        path: &str, period_ms: u128, max_bytes: usize,
+    ) -> std::result::Result<Self, std::io::Error> {
+        let random_file = File::open(path)?;
+
+        Ok(VuRngBackend {
+            mem: None,
+            event_idx: false,
+            random_file: Mutex::new(random_file),
+            timer: Mutex::new(VuRngTimerConfig::new(period_ms, max_bytes)),
+            exit_event: EventFd::new(EFD_NONBLOCK).expect("Creating exit eventfd"),
+        })
+    }
+
+    pub fn process_queue(&self, vring: &mut Vring) -> Result<bool> {
+        while let Some(mut desc_chain) = vring.mut_queue().iter().unwrap().next() {
+            let head_index = desc_chain.head_index();
+            let descriptor = desc_chain.next().unwrap();
```

Comment:
Don't use unwrap() :)

Reply:
To be clear - did you want to see all the unwrap() removed or just the ones you pointed out?